### ISSUE TITLE
Option for interactive control of Window Width and Window Level via mouse click-and-drag 

### ISSUE THREAD

**Issue by JDWarner** [napari/napari#3388]
## 🚀 Feature
Radiologists expect interactive control of the "window width and window level".  Napari does not currently expose this, but easily could.  This reframes the concept currently expressed in Napari on an image layer by "contrast limits".  Instead of two variables representing `min` and `max`, this would become the two variables `window_width` and `window_level` defined simply as the total width of the window and its midpoint.  These would be defined from the current max and min as
`window_width = max - min` and 
`window_level = (max - min) / 2`

or in the opposite direction
`max = window_level + window_width / 2`
`min = window_level - window_width / 2`

The reason the window width and window level concept is used is that it much more naturally allows interactive adjustment via the mouse.  One of these variables is mapped to the vertical axis, the other mapped to the horizontal axis; this generally can be swapped in settings and one or both can be reversed per user preference.  Mapping `min` and `max` to these dimensions directly is not an intuitive way to explore these data.

The current state of these variables should also be at least optionally overlaid in the corner of the viewer as, for example, a soft tissue CT window would show
```
ww: 440
wl: 40
```

## Motivation
This allows efficient dynamic exploration of high bit rate image data, which requires mapping from full range (e.g., 65k+ values for a 16 bit image) to the 256 values most monitors can display.  Such data is quite common in medical imaging.  Radiologists have trained their eye to recognize the tissue/data of interest as mid-gray when the level is correct, and then "zoom" the width as appropriate.  This operation happens generally in 0-2 seconds, and is repeated frequently depending on the anatomy of interest (soft tissues vs. lung vs. bone).

## Pitch
Napari already exposes a rich interface akin to a miniature PACS viewer.  Radiologists _expect_ the ability to interactively adjust the viewed portion of the histogram by window width and level, which is a standard feature across all PACS systems.  This should be simple to implement in Napari and will make the viewer immediately more functional for dynamic exploration of higher bit depth images and tasks like medical image annotation.

## Alternatives
The current contrast min/max is clunky, not interactive, and not intuitive to the user for interactive exploration.  It _can_ be coerced to show the data in an equivalent manner but requires much more time and effort to interact with instead of fluid exploration.  The slider is also small, thus imprecise, and the exact values of min and max are not shown.

## Additional context
I'm not familiar with Napari's interface and GUI signaling, so it isn't clear if this could be implemented through the plugin infrastructure or needs deeper modification.  With some guidance on path forward, I may be able to help contribute this feature.

**Comment by tlambert03** [comment]
fascinating :) thanks for the very clear issue @JDWarner!
in the domain of microscopy (where a lot of our core devs come from ... but which we definitely don't want to be tied to!) min/max is a common way to express this, since it deals directly with the domain of numbers in the raw dataset (i.e. we generally know exactly what it means about the signal on the camera when our max brightness corresponds to a particular gray value in an N bit range).

this sounds much more like a "brightness and contrast" way to adjust things (where brightness is the offset, or your "level", and contrast is the slope of the lookup, or your "width").  It seems likely that there will be other domains that are more familiar with width/level, or brightness/contrast than min/max, so it does seem like a good thing to offer.

we should gather some opinions as to whether this should a feature where the user always sees both ways of expressing the LUT, or whether it should be a configurable settings option.

As for the implementation: this wouldn't be easy to do as a plugin... it needs deeper modification.  steps needed:

- [ ] decide whether this is a GUI-only concept or a programmatic one
  - [ ] if programmatic control is needed, add some programmatic access in https://github.com/napari/napari/blob/main/napari/layers/intensity_mixin.py that converts between ww/wl and min/max
- [ ] for the display, we'd need to modify the contrast limits slider [here](https://github.com/napari/napari/blob/24d01f5fee450c881c5f8fce4d17395f9af2328e/napari/_qt/layer_controls/qt_image_controls_base.py#L91) to be "mode-aware" (depending on whether the user wanted to control ww/wl or min/max)
- [ ] the ww/wl text overlay could be added either as text to the status bar (at the bottom of the window), or as an optional [text overlay](https://github.com/napari/napari/blob/main/napari/components/text_overlay.py) in the canvas itself.  (also, regardless of whether we're in min/max or ww/wl, we need better numeric display of the contrast limits)

**Comment by JDWarner** [comment]
The ability to interface programmatically would be ideal for uses like a plugin with keybindings or a drop down box presenting several standard named presets; in the setting of CT data this might include 'Brain', 'Bone', 'Soft Tissue', 'Lung', and 'Liver'.

The standard in this field would be a small text corner image overlay, but so long as they can be shown _somewhere_ the spirit of the request would be met.  At present I can play with the slider but not see what the min and max are set to - in both modes I think it would be valuable to show the values somewhere unobtrusive.

**Comment by tlambert03** [comment]
> At present I can play with the slider but not see what the min and max are set to

yep this is an issue we're well aware of, discussions elsewhere.  you can right click on the slider to get a lot more detail

**Comment by jni** [comment]
Thanks for the issue @JDWarner, and :wave:!

@tlambert03 one hacky-but-fast way to implement this is to add a scroll callback to the dual sliders to control the width, since the center of the slider is already draggable (to control the level). We could perhaps even make up/down drag control the width, though personally I would find that weird.

Adding a text overlay programmatically is easy @JDWarner (if poorly documented), see [this example](https://github.com/napari/napari/blob/24d01f5fee450c881c5f8fce4d17395f9af2328e/examples/viewer_fps_label.py).

@JDWarner do you have a screenshot for what the adjustment widget looks like on other software? I'm intrigued by the two-axis mouse control and whether it still looks like a horizontal bar or more like a point in a 2D square? Or an isosceles triangle perhaps (since the center can't move at max width)?

**Comment by tlambert03** [comment]
> add a scroll callback to the dual sliders to control the width

Actually that's already implemented.  If you hold down option/alt while scrolling over that slider, it will contract/broaden.  (If it's not option it's another key, I need to check)

**Comment by jni** [comment]
A-hah! It *is* option! Funny I had this memory but I couldn't make it work with just plain scrolling. LOL Should we get rid of the modifier?

**Comment by tlambert03** [comment]
but without the modifier it should scroll both of them, right?  so with modifier is width, without modifier is level

**Comment by jni** [comment]
no, without modifier it does jack. :joy: To move the level you click and drag in the middle of the bar.

**Comment by tlambert03** [comment]
oh sorry... just checked, hold command for offset/level:


https://user-images.githubusercontent.com/1609449/133613757-06e0d0ab-0cc4-415f-a902-a15fbb043659.mov

**Comment by jni** [comment]
My earlier question still stands: should we remove the need to use option for the width? I think yes. Then everything can be controlled with the mouse, no keyboard needed.

**Comment by tlambert03** [comment]
sure, happy to change up the modifiers.  It feels slightly unusual to me have it control width without a modifier (like, it's inconsistent with other slider behavior, and a little surprising), but I get your point that it's the only way to remove the need for the keyboard (since we can click/drag the bar for level).

**Comment by JDWarner** [comment]
@jni (👋 !) there is no widget, essentially all PACS systems map this to two axes of mouse movement as described. The window and level are independent in each viewpane and there's really no straightforward method to have 12+ 2D widgets.  The only visual confirmation is the noted ww/wl variables usually somewhere in the corner of the view.

I have been experimenting with some Jupyter widgets which overlay the window and level on the image histogram like this (soft tissue window on a CT shown) for teaching, but Napari doesn't natively show the histogram - not sure the sidebar has enough width to do that well. 
![image](https://user-images.githubusercontent.com/2184487/133631782-7c118d96-0fb5-4995-babb-b8763e6e7ba2.png)


To be more specific about the mapping, generally 
* Vertical movement is mapped to level and inverted, so moving toward top of screen actually moves the center of the window to lower values.  This is because doing so makes the image brighter, so it 'feels' correct even though it's actually inverted.
* Horizontal movement is mapped to width and not inverted; moving right produces a wider window and moving left a narrower one.

That's the mapping I have seen most often and seems to be the default for web-based PACS and image viewers which are less customizable.  Most standalone PACS systems do allow these to be swapped and inverted though, so I assume there are people with strong alternative preferences out there.

**Comment by haesleinhuepf** [comment]
Hi @JDWarner ,

I just found this issue. I've implemented this feature as part of a napari-plugin:
https://github.com/haesleinhuepf/napari-mouse-controls#windowing

Feedback welcome!

Best,
Robert
